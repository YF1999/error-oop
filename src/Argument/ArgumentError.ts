import { IErrorArguments, IErrorOptions, NativeError } from '../Native';
import { ErrorTool } from '../Tools';

export interface IArgumentErrorOptions extends IErrorOptions {
    /**
     * Do not append message suffix at the end of original message.
     */
    noMessageSuffix?: boolean;
    /**
     * The name of the parameter that caused the current error.
     */
    paramName?: string;
}

/**
 * Applicable when one of the arguments provided to a function or method is not valid.
 */
export class ArgumentError extends NativeError {
    protected _paramName?: string;

    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be appended.
     */
    public constructor(message: string, innerError: Error);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     */
    public constructor(message: string, paramName: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param innerError The error that is the cause of the current error. Stack trace will be appended.
     */
    public constructor(message: string, paramName: string, innerError: Error);
    /**
     * @param options The constructor options.
     */
    public constructor(options: IArgumentErrorOptions);

    public constructor(...args: IErrorArguments<IArgumentErrorOptions, ['paramName']>) {
        const options = ErrorTool.parseErrorArguments(['paramName'], ...args);

        if (options.paramName && options.noMessageSuffix !== true) {
            options.message = `${options.message} (Parameter '${options.paramName}')`;
        }

        super(options);

        this._paramName = options.paramName;
        this._setNonEnumerable('_paramName');
    }

    public get paramName(): string | undefined {
        return this._paramName;
    }
}
