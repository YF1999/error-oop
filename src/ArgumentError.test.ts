import { ArgumentError } from './ArgumentError';

describe('Test `ArgumentError`', () => {
    const name = 'ArgumentError';
    const message = 'error';
    const paramName = 'key_parameter';

    function getErrors() {
        const err1 = new ArgumentError();
        const err2 = new ArgumentError(message);
        const err3 = new ArgumentError(message, err2);
        const err4 = new ArgumentError(message, paramName, err3);
        const err5 = new ArgumentError({ message, innerError: err4 });
        const err6 = new ArgumentError({ message, innerError: err5, paramName });
        const err7 = new ArgumentError({ message, innerError: err6, noMessageSuffix: true });
        const err8 = new ArgumentError({ message, innerError: err7, paramName, noMessageSuffix: true });

        return [err1, err2, err3, err4, err5, err6, err7, err8];
    }

    test('property: error name', () => {
        getErrors().map((err) => expect(err.name).toBe(name));
    });

    test('property: message', () => {
        const [err1, ...errs] = getErrors();
        expect(err1.message).toBe('');
        errs.map((err) => expect(err.message).toMatchSnapshot());
    });

    test('property: inner error', () => {
        const [err1, err2, err3, err4, err5, err6, err7, err8] = getErrors();
        expect(err1.innerError).toBeUndefined();
        expect(err2.innerError).toBeUndefined();
        expect(err3.innerError).toBe(err2);
        expect(err4.innerError).toBe(err3);
        expect(err5.innerError).toBe(err4);
        expect(err6.innerError).toBe(err5);
        expect(err7.innerError).toBe(err6);
        expect(err8.innerError).toBe(err7);
    });

    test('property: parameter name', () => {
        const [err1, err2, err3, err4, err5, err6, err7, err8] = getErrors();
        expect(err1.paramName).toBeUndefined();
        expect(err2.paramName).toBeUndefined();
        expect(err3.paramName).toBeUndefined();
        expect(err4.paramName).toBe(paramName);
        expect(err5.paramName).toBeUndefined();
        expect(err6.paramName).toBe(paramName);
        expect(err7.paramName).toBeUndefined();
        expect(err8.paramName).toBe(paramName);
    });

    test('empty serialization', () => {
        getErrors().map((err) => expect(JSON.stringify(err)).toBe('{}'));
    });

    test('no enumerable fields', () => {
        getErrors().map((err) => expect(Object.keys(err).length).toBe(0));
    });

    test('own keys', () => {
        const [err1, ...errs] = getErrors();
        expect(Reflect.ownKeys(err1)).toStrictEqual(['stack', '_paramName']);
        errs.map((err) => expect(Reflect.ownKeys(err)).toStrictEqual(['stack', 'message', '_paramName']));
    });
});
