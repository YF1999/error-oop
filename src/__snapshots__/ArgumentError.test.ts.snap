// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`ArgumentError\` property: message 1`] = `"error"`;

exports[`Test \`ArgumentError\` property: message 2`] = `"error"`;

exports[`Test \`ArgumentError\` property: message 3`] = `"error (Parameter 'key_parameter')"`;

exports[`Test \`ArgumentError\` property: message 4`] = `"error"`;

exports[`Test \`ArgumentError\` property: message 5`] = `"error (Parameter 'key_parameter')"`;

exports[`Test \`ArgumentError\` property: message 6`] = `"error"`;

exports[`Test \`ArgumentError\` property: message 7`] = `"error"`;
